jms/org.wso2.carbon.transport.jms/src/main/java/org/wso2/carbon/transport/jms/factory/CachedJMSConnectionFactory.java

```diff
@@ -0,0 +1,323 @@
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public MessageProducer getMessageProducer(Session session, Destination destination)
+            throws JMSConnectorException {
+        MessageProducer messageProducer;
+        if (cachedMessageConsumer == null) {
+            messageProducer = createMessageProducer(session, destination);
+        } else {
+            messageProducer = cachedMessageProducer;
+        }
+        return messageProducer;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public MessageProducer createMessageProducer(Session session, Destination destination)
+            throws JMSConnectorException {
+        MessageProducer messageProducer = super.createMessageProducer(session, destination);
+        if (this.cacheLevel >= JMSConstants.CACHE_PRODUCER) {
+            cachedMessageProducer = messageProducer;
+        }
+        return messageProducer;
+    }
+
+    public void closeConnection() throws JMSException {
+        if (cachedConnection != null) {
+            cachedConnection.close();
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void closeConnection(Connection connection) throws JMSConnectorException {
+        try {
+            if (this.cacheLevel < JMSConstants.CACHE_CONNECTION) {
+                connection.close();
+            }
+        } catch (JMSException e) {
+            throw new JMSConnectorException("JMS Exception while closing the connection.", e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void closeMessageConsumer(MessageConsumer messageConsumer) throws JMSConnectorException {
+        try {
+            if (this.cacheLevel < JMSConstants.CACHE_CONSUMER) {
+                messageConsumer.close();
+            }
+        } catch (JMSException e) {
+            throw new JMSConnectorException("JMS Exception while closing the consumer.", e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void closeMessageProducer(MessageProducer messageProducer) throws JMSConnectorException {
+        try {
+            if (this.cacheLevel < JMSConstants.CACHE_PRODUCER) {
+                messageProducer.close();
+            }
+        } catch (JMSException e) {
+            throw new JMSConnectorException("JMS Exception while closing the producer.", e);
+        }
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void closeSession(Session session) throws JMSConnectorException {
+        try {
+            if (this.cacheLevel < JMSConstants.CACHE_SESSION) {
+                session.close();
+            }
+        } catch (JMSException e) {
+            throw new JMSConnectorException("JMS Exception while closing the session.", e);
+        }
+    }
+
+    private void resetCache() throws JMSException {
+        if (cachedMessageConsumer != null) {
+            try {
+                cachedMessageConsumer.close();
```

Comment:
Do we need the try catch blocks here? We have the JMSException in the method signature and we are not doing anything special in the catch block.

Reply:
These are fixed.